src/qibo/tomography/gate_set_tomography.py

```diff
@@ -0,0 +1,277 @@
+        For two qubits, \\(k \\in \\{0, 1, 2, 3\\}^{\\otimes 2}\\).
+        nqubits (int): Number of qubits.
+
+    Returns:
+        list(:class:`qibo.gates.abstrac.Gate`): list of the gates that prepare the k-th state.
+    """
+
+    if not nqubits in (1, 2):
+        raise_error(
+            ValueError,
+            f"nqubits needs to be either 1 or 2, but is {nqubits}.",
+        )
+    gates = _gates(nqubits)[k]
+    return [gate(q) for q in range(len(gates)) for gate in gates[q]]
+
+
+@cache
+def _measurement_basis(j, nqubits):
+    r"""Constructs the j-th measurement basis for a `nqubits`-gate.
+
+        Args:
+        j (int): The index of the measurement basis.
+            For a single qubit, \(j \in \{0, 1, 2, 3\} \equiv \{I, X, Y, Z\}\).
+            For two qubits, \(j \in \{0, 1, 2, 3\}^{\otimes 2}\).
+        nqubits (int): Number of qubits..
+
+    Returns:
+        List[:class:`qibo.gates.abstrac.Gate`]: list of gates forming the :math:`j`-th element of the Pauli measurement basis.
+    """
+
+    if not nqubits in (1, 2):
+        raise_error(
+            ValueError,
+            f"nqubits given as {nqubits}. nqubits needs to be either 1 or 2.",
+        )
+
+    measurements = _measurements(nqubits)[j]
+    return [gates.M(q, basis=measurements[q]) for q in range(len(measurements))]
+
+
+def reset_register(circuit, invert_register):
+    """Returns an inverse circuit of the selected register to prepare the zero state \\(|0\rangle\\).
+        One can then add inverse_circuit to the original circuit by addition:
+            circ_with_inverse = circ.copy()
+            circ_with_inverse.add(inverse_circuit.on_qubits(invert_register))
+        where register_to_reset = (0,), (1,) , or (0, 1).
+
+        Args:
+        circuit (:class:`qibo.models.Circuit`): original circuit
+        invert_register (tuple): Qubit(s) to reset: Use a tuple to specify which qubit(s) to reset:
+            - (0,) to reset qubit 0;
+            - (1,) to reset qubit 1; or
+            - (0,1) to reset both qubits.
+    Returns:
+        inverse_circuit (:class:`qibo.models.Circuit`): Inverse of the input circuit's register.
+    """
+    valid_registers = [(0,), (1,), (0, 1)]
+    if invert_register is not None:
+        if (
+            not isinstance(invert_register, tuple)
+            or invert_register not in valid_registers
+        ):
+            raise_error(
+                NameError,
+                f"Invalid register {invert_register}, please pick one in {valid_registers}.",
+            )
+
+        elif invert_register == (0,) or invert_register == (1,):
+            register_to_reset = invert_register[0]
+            new_circ = Circuit(1)
+            for data in circuit.raw["queue"]:
+                init_kwargs = data.get("init_kwargs", {})
+                if data["_target_qubits"][0] == register_to_reset:
+                    new_circ.add(getattr(gates, data["_class"])(0, **init_kwargs))
```

Comment:
```suggestion
            for gate in circuit.queue:
                if register_to_reset in gate.target_qubits:
                    new_circ.add(gate.__class__(0, **gate.init_kwargs))
```
I would directly use the `gate` object, but still I am not completely sure what you are doing here. 
I understand, that you are creating a one qubit circuit containing all the gates that act on the qubit you want to reset, then you invert it and append it to the original circuit on the corresponding qubit, but I have two concerns. How does this work with two-qubits gates? For example, if you have a `CZ(0,1)` and you want to invert register `1` you will have `CZ.target_qubits[0]==1`, but then building the new gate `CZ(0, **init_kwargs)` will fail, right? More generally, instead, I am not sure what you get out of GST when you reset a register, since you're effectively changing the state preparation I would expect to obtain something different, right?

Reply:
I think using the `gate` object is straightforward and easy to implement. I'm not sure how it will work with `CZ`. But resetting 2 qubit gates is not straightforward.

Generally, resetting the register when the state preparation has already been initialized is ok since we don't have 2 qubit gates in state preparation. We're doing just that under the function `_gate_tomography()` (which I've amended the name of the function to `_gate_set_tomography()` ). 

I will make a note that this cannot be used for entangling gates. Perhaps we can make arrangements for using ancillary qubits to reset the qubit of interest when we do PEC.

Reply:
The resetting of the register is only used for basis operations which call for a fresh state to be prepared, like the ones stated in [Physical Review Research 3.3 (2021): 033178](https://journals.aps.org/prresearch/abstract/10.1103/PhysRevResearch.3.033178).  I have included a note inside the documentation.

Reply:
Ok, so, since this seems to be related to `PEC` and only indirectly to `GST`, I would suggest to remove this from this PR and include it in the next one, when we are going to actually deal with `PEC`. At this stage we don't have any proper way of testing this from what I see.

Reply:
Sure, we can move the `reset_register` to `PEC`. Agree, it will help with testing. 

Anyway, just a side note, I will be on course next week so apologies in advance if there are any further delays. Will try to get it done as soon as I can.